patches/tock/13-send-packet-before-receiving.patch

```diff
@@ -0,0 +1,32 @@
+diff --git a/capsules/src/usb/usb_ctap.rs b/capsules/src/usb/usb_ctap.rs
+index 2c91c0968..ea8111069 100644
+--- a/capsules/src/usb/usb_ctap.rs
++++ b/capsules/src/usb/usb_ctap.rs
+@@ -253,18 +253,19 @@ impl<'a, 'b, C: hil::usb::UsbController<'a>> Driver for CtapUsbSyscallDriver<'a,
+                             if app.waiting {
+                                 ReturnCode::EALREADY
+                             } else {
++                                // Indicates to the driver that we have a packet to send.
++                                let r = self
++                                    .usb_client
++                                    .transmit_packet(app.buffer.as_ref().unwrap().as_ref(), endpoint);
```

Comment:
Currently the code is like:
```
self.received_packet()
if app.busy {
  return ReturnCode::SUCCESS
}
self.send_packet(buf)
```

So if the receive_packet() call picks up a packet and starts processing it, the app will be busy, this function will return success, but the packet was not sent. From the perspective of the caller, it called send_or_recv() and it returned success, so it has to assume that the packet was sent. Therefore the packet is never sent and next call the next packet will be sent. I think this is clear that this is an issue.

Correct, by sending first, we can't respond differently based on a received packet. However, that is how things currently work, because the sending of replies happens in a tight loop in main.rs without processing any new packets.
I'm not sure the expectations of the host, but expecting a new packet to change the behavior of an already processed packet is inherently racey, so I'm not too concerned about not responded to the received packet promptly. 

But I'm open to other thoughts.

Reply:
I see, the main problem is that the API is not able to differentiate between:
- successful send and receive
- successful send
- successful receive

I think the previous convention (because of the name "send OR receive") was that "successful send and receive" never happens. The last two are distinguished using some return value? (I don't know)

I agree with you that the current code is wrong because if `send_or_recv` receives, it doesn't try to send again the packet and instead sends the next one (if any). Which suggests that `send_or_recv` is used as if it was `send_and_maybe_recv`. And this is what you're actually doing in this PR (fixing the implementation to reflect the usage and not the name).

I guess renaming is a pain so maybe just documenting `api::HidConnection::send_or_recv_with_timeout` with the expected semantics is already good enough. We could also possibly just rename that function and keep the Tock naming wrong. (I actually prefer that.)

Reply:
yes, exactly.

I've renamed send_or_recv to send_and_maybe_recv.